File: C#/src/Core/Logger.cs
using System.Text.Json;

namespace Csharp.Core;

public static class Logger
{
    private static readonly string LogFile = Path.Combine(
        Directory.GetCurrentDirectory(),
        "storage",
        "logs",
        "app.log"
    );

    public static void Info(string message, Dictionary<string, object>? meta = null, Dictionary<string, object>? data = null)
    {
        Write("INFO", message, meta, data);
    }

    public static void Error(string message, Dictionary<string, object>? meta = null, Dictionary<string, object>? data = null)
    {
        Write("ERROR", message, meta, data);
    }

    private static void Write(string level, string message, Dictionary<string, object>? meta, Dictionary<string, object>? data)
    {
        var log = new
        {
            timestamp = DateTime.Now.ToString("yyyyMMddTHH:mm:sszzz"),
            message,
            level,
            trace_id = GenerateTraceId(),
            service = "users-api",
            meta_data = meta ?? new Dictionary<string, object>(),
            data = data ?? new Dictionary<string, object>(),
        };

        var json = JsonSerializer.Serialize(log);

        Directory.CreateDirectory(Path.GetDirectoryName(LogFile)!);

        File.AppendAllText(LogFile, json + Environment.NewLine);
    }

    private static string GenerateTraceId()
    {
        return $"{Guid.NewGuid():N},{Guid.NewGuid():N}";
    }
}

File: dart/lib/core/logger.dart
import 'dart:convert';
import 'dart:io';

class Logger {
  static final String logFile = 'storage/logs/app.log';

  static void info(String message,
      {Map<String, dynamic>? metaData, Map<String, dynamic>? data}) {
    _write("INFO", message, metaData, data);
  }

  static void error(String message,
      {Map<String, dynamic>? metaData, Map<String, dynamic>? data}) {
    _write("ERROR", message, metaData, data);
  }

  static void _write(String level, String message,
      Map<String, dynamic>? metaData, Map<String, dynamic>? data) {
    final log = {
      "timestamp": DateTime.now().toIso8601String(),
      "message": message,
      "level": level,
      "trace_id": _generateTraceId(),
      "service": "users-api",
      "meta_data": metaData ?? {},
      "data": data ?? {},
    };

    final file = File(logFile);
    file.createSync(recursive: true);
    file.writeAsStringSync(jsonEncode(log) + "\n",
        mode: FileMode.append);
  }

  static String _generateTraceId() {
    return "${_randomHex()},${_randomHex()}";
  }

  static String _randomHex() {
    return DateTime.now().microsecondsSinceEpoch.toRadixString(16);
  }
}